monitor/src/main.rs

```diff
@@ -0,0 +1,167 @@
+mod config;
+mod controller;
+mod external_api;
+mod event_listener;
+
+use crate::config::{Config, SingleConfig};
+use crate::event_listener::{get_github_repo_url, listen_to_commits, listen_to_pull_requests};
+use crate::external_api::{add_configuration, delete_configuration, get_actions_file, get_configuration_by_id, get_configurations, update_configuration, AppState};
+use actix_web::{web, App, HttpServer};
+use clap::{Arg, Command};
+use controller::send_to_controller;
+use tokio::task::JoinHandle;
+use std::path::Path;
+use std::sync::Arc;
+use tokio;
+use tokio::sync::RwLock;
+
+#[tokio::main]
+async fn main() -> std::io::Result<()> {
+    let config = get_config();
+    let configs = Arc::new(RwLock::new(config));
+
+    let mut thread_listeners_handles: Vec<JoinHandle<()>> = vec![];
+
+    println!("Launching API and listening to events on Github repository...");
+
+    tokio::select! {
+        _ = launch_api_server(Arc::clone(&configs)) => {},
+        _ = launch_github_listeners(Arc::clone(&configs), &mut thread_listeners_handles) => {},
+    }
+
+    println!("Threads launched, waiting for them to finish...");
+    println!("{:#?}", thread_listeners_handles);
+
+    futures::future::join_all(thread_listeners_handles).await;
+
+    Ok(())
+}
+
+fn get_config() -> Config {
+    // CLI arguments
+    let matches = Command::new("GitHub Monitor")
+        .version("1.0")
+        .about("Monitors a GitHub repository for changes")
+        .arg(Arg::new("config")
```

Comment:
All args of your cli are not required even thought they look mandatory. You should put default values where you can and maybe make some field required

Reply:
they are not mandatory here because the check of this is just after. Hence, the user can either pass just a config file, or just every necessary arguments directly (repo_name, repo_user ...), if none of these are precised by the user, then an error is thrown. However if just the config file is precised and nothing else, for example, no error is thrown, because every information about the repo is in the config file, not in the arguments. It is either one or the other, or both, and no clap option allows to do exactly a "required" like that.